tests/test_graphofgrid.cpp

```diff
@@ -0,0 +1,517 @@
+#include <opm/input/eclipse/Schedule/Well/WellConnections.hpp>
+#include <opm/input/eclipse/Schedule/Well/Well.hpp>
+
+// basic test to check if the graph was constructed correctly
+BOOST_AUTO_TEST_CASE(SimpleGraph)
+{
+    Dune::CpGrid grid;
+    std::array<int,3> dims{2,2,2};
+    std::array<double,3> size{2.,2.,2.};
+    grid.createCartesian(dims,size);
+    Opm::GraphOfGrid gog(grid);
+
+    BOOST_REQUIRE(gog.size()==8); // number of graph vertices
+    BOOST_REQUIRE(gog.numEdges(0)==3); // each vertex has 3 neighbors
+
+    auto edgeL = gog.edgeList(2);
+    BOOST_REQUIRE(edgeL.size()==3); // neighbors of vertex 2 are: 0, 3, 6
+    BOOST_REQUIRE(edgeL[0]==1.);
+    BOOST_REQUIRE(edgeL[3]==1.);
+    BOOST_REQUIRE(edgeL[6]==1.);
+    BOOST_REQUIRE_THROW(edgeL.at(4),std::out_of_range); // not a neighbor (edgeL's size increased)
+
+    BOOST_REQUIRE_THROW(gog.edgeList(10),std::logic_error); // vertex 10 is not in the graph
+}
+
+// test vertex contraction on a simple graph
+BOOST_AUTO_TEST_CASE(SimpleGraphWithVertexContraction)
+{
+    Dune::CpGrid grid;
+    std::array<int,3> dims{2,2,2};
+    std::array<double,3> size{2.,2.,2.};
+    grid.createCartesian(dims,size);
+    Opm::GraphOfGrid gog(grid);
+
+    auto edgeL = gog.edgeList(3); // std::map<int,float>(gID,edgeWeight)
+    BOOST_REQUIRE(edgeL[1]==1);
+    BOOST_REQUIRE_THROW(edgeL.at(0),std::out_of_range);
+    gog.contractVertices(0,1);
+    BOOST_REQUIRE(gog.size()==7);
+    edgeL = gog.edgeList(3);
+    BOOST_REQUIRE_THROW(edgeL.at(1),std::out_of_range);
+    BOOST_REQUIRE(edgeL[0]==1);
+    edgeL = gog.edgeList(0);
+    BOOST_REQUIRE(edgeL.size()==4);
+    BOOST_REQUIRE(edgeL[2]==1); // neighbor of 0
+    BOOST_REQUIRE(edgeL[3]==1); // neighbor of 1
+    BOOST_REQUIRE_THROW(edgeL.at(1),std::out_of_range); // removed vertex, former neighbor of 0
+
+    gog.contractVertices(0,2);
+    BOOST_REQUIRE(gog.size()==6);
+    BOOST_REQUIRE(gog.getVertex(0).weight==3.);
+    edgeL = gog.edgeList(0);
+    BOOST_REQUIRE(edgeL.size()==4);
+    BOOST_REQUIRE(edgeL[3]==2);
+    BOOST_REQUIRE(gog.edgeList(3).size()==2);
+    BOOST_REQUIRE(gog.edgeList(3).at(0)==2);
+    // contracting vertices removes higher ID from the graph
+    // (when well is added, IDs removed from the graph are stored in the well)
+    BOOST_REQUIRE_THROW(gog.getVertex(1),std::logic_error);
+
+    auto v5e = gog.getVertex(5).edges;
+    BOOST_REQUIRE(v5e==gog.edgeList(5));
+    BOOST_REQUIRE(v5e==gog.edgeList(6)); // 5 and 6 have the same neighbors (1, 2 got merged)
+    BOOST_REQUIRE(v5e!=gog.edgeList(7));
+
+}
+
+BOOST_AUTO_TEST_CASE(WrapperForZoltan)
+{
+    Dune::CpGrid grid;
+    std::array<int,3> dims{5,4,3};
+    std::array<double,3> size{1.,1.,1.};
+    grid.createCartesian(dims,size);
+    Opm::GraphOfGrid gog(grid);
+
+    int err;
+    int nVer = getGraphOfGridNumVertices(&gog,&err);
+    BOOST_REQUIRE(err==ZOLTAN_OK);
+    BOOST_REQUIRE(nVer == 60);
+
+    int gIDs[nVer], lIDs[0];
```

Comment:
I get many warnings now from this and other places in the code: arrays with variable length is not C++, it is a GNU C extension, so we should not use it. Use vector instead:
```
std::vector<int> gIDs(nVer);
```
And then if you need to pass the pointer to the start of the data use `gIDs.data()`.

However, the `lIDs[0]` has me confused: a zero length array? Why? In general, it seems to me this should have the same size as the other array. For getGraphOfGridVerticesList() specifically, I think it does not look at that array at all (it is flagged as maybe_unused), then passing `nullptr` here is appropriate (although I think the arrays, sizes and content should be similar to how it will be used in real code if possible).

@michal-toth can you address these issues?

Reply:
Thanks, I was not aware of the std::vector's .data() function. I cleared it out, hopefully I found all instances.